rfcs/015 - TCP RFC.md

```diff
@@ -0,0 +1,196 @@
+- Colton Donnely <colton@alantechnologies.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- David Ellis <david@alantechnologies.com>
+
+## Summary
+
+The current `@std/http` in Alan is perfectly serviceable for "normal" web servers that consume a request in its entirety, then provide a response in its entirety, and then close the connection. But if the server was to drip-feed data back to the client, or wanted to upgrade into a two-way websocket connection, this is impossible. Adding support for arbitrary TCP messaging to Alan solves this for pass-through and provides a foundation for a fully-Alan HTTP implementation. It also stretches the language with its static event system, but not to the breaking point.
+
+## Expected SemBDD Impact
+
+This would be a minor update if Alan was post-1.0.0 as it adds new functionality without regressing on existing functionality.
+
+## Proposal
+
+A TCP connection has four essential states: open, closed-on-purpose, closed-on-error, and closed-by-timeout. While in the open state, data can be read from its output, written to its input, or explicitly closed as triggers by the user, while it can also unexpectedly transition to the other two closed states. The amount of data received at any particular point in time is completely arbitrary and up to both the originating machine and the local kernel, which makes it a good fit for the event system so the users' code is prompted when each chunk is ready to work with that it can continue with.
+
+The main issue is that Alan's event system is intentionally a static event system determined at compile time, so it is impossible to add new event handlers for each new socket that is opened by the users' code or has been received by the TCP server. The secondary issue is that the AVM's internal memory model would require an expensive conversion of that chunk from packed bytes to an internal `Array<int8>` type which, to simplify the memory model, is actually really an `Array<int64>` under the hood, which is very wasteful for a true byte array, so being able to avoid that conversion when possible would be good. Finally, the ergonomics of the API need to be both easy-to-understand and mesh well with the existing syntax while adding stream processing support at the same time.
+
+While changes to the language are on the table (and some of the alternatives will cover a few of them), the best overall solution (considering implementation complexity and cognitive complexity) doesn't need it. :)
+
+The clearest way to introduce the new syntax is to simply use it in an example similar to how it will be used by Anycloud:
+
+```ln
+from @std/tcpserver import connection, Connection
+from @std/tcp import connect, addContext, chunk, Chunk, Context, read, write, connClose, close, connError, connTimeout
```

Comment:
is there a way to consolidate the imports or make them more intuitive / less error-prone? It seems like `connect` returns a `connection` that one can call `addContext` on, but `connection` belongs to `tcpserver` unlike the rest which belong to `tcp`. also it might be more clear to just rename to `tcpclient` and `tcpserver`

Reply:
The only thing that needs to be imported from `@std/tcpserver` is the actual `connection` event, not even the `Connection` type, which could also be from the regular `@std/tcp` but I thought that looked a bit weird? Would it be less weird to you if everything is imported from `@std/tcp` and only the `connection` event is isolated in `@std/tcpserver`?

Reply:
That would be seem less weird to me